src/hal/link/desktop/serial-link.cpp

```diff
@@ -0,0 +1,133 @@
+        );
+
+        // We should remove flow control to make sure that will work in any case
+        _serialPort.set_option(boost::asio::serial_port::flow_control(
+            boost::asio::serial_port::flow_control::none));
+    } catch (const std::exception& exception) {
+        std::cerr << "Error while configuring serial port: " << exception.what() << std::endl;
+        close();
+        return;
+    }
+
+    // Bind event loop start with the async read schedulement
+    _context.eventLoop.post(boost::bind(&SerialLink::bindRead, this));
+
+    // Allow event loop to run, for now
+    _context.run = true;
+    _context.future = std::async(std::launch::async, [this] {
+        // Run event loop
+        while (_context.run) {
+            using namespace std::chrono_literals;
+            try {
+                _context.eventLoop.run_for(100ms);
+            } catch (const std::exception& exception) {
+                std::cerr << "Event loop failed to run: " << exception.what() << std::endl;
+            }
+        }
+        _context.eventLoop.stop();
+    });
+
+    // Start Automatic Baud Rate procedure
+    _serialPort.send_break();
+    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+    _serialPort.write_some(boost::asio::buffer("UUUUUUUUUU"));
+}
+
+SerialLink::~SerialLink()
+{
+    close();
+
+    // Ask context to finish and wait for it
+    _context.run = false;
+    _context.future.wait();
+}
+
+void SerialLink::bindRead()
+{
+    _serialPort.async_read_some(
+        boost::asio::buffer(_rxBuffer),
+        boost::bind(&SerialLink::doRead, this, boost::asio::placeholders::error,
+            boost::asio::placeholders::bytes_transferred));
+}
+
+void SerialLink::doRead(boost::system::error_code error, size_t bytesReceived)
+{
+    if (error) {
+        // The read can canceled by someone else or when class is deleted
+        if (error.value() == boost::system::errc::operation_canceled) {
+            return;
+        }
+        // That's sad, something is really wrong!
+        std::cerr << "Error while reading from serial port: " << error.message() << std::endl;
+        // Close the serial port and abort
+        if (_serialPort.is_open()) {
+            _serialPort.close();
+        }
+    } else {
+        // We got something
+        if (bytesReceived) {
+            // Notify signal that we got something
+            const std::vector<uint8_t> output(std::cbegin(_rxBuffer), std::next(std::cbegin(_rxBuffer), bytesReceived));
+            // Emit signal
+            _onReceived(output);
+
+            // Update our shared buffer
+            _linkBuffer.mutex.lock();
+            _linkBuffer.data.insert(std::end(_linkBuffer.data), std::begin(_rxBuffer), std::next(std::begin(_rxBuffer), bytesReceived));
+            _linkBuffer.mutex.unlock();
+        }
+
+        // Schedule a new async read
+        bindRead();
+    }
+}
+
+int SerialLink::read(uint8_t* buffer, int nBytes)
+{
+    _linkBuffer.mutex.lock();
```

Comment:
use a mutex locker

Reply:
That's a mutex locker, I don't know how `mutex.lock` can't be clear.

Reply:
If you are talking about scope mutex lockers, this code is only used in two places. It'll be also be necessary to add another variable or a variable and a function inside the struct, this will also in a bigger scope for the RAII locker logic, so there is not a ruch win here for now, maybe in the future if we use the lockers more.